### PR TITLE
[GPU] Weightless caching - propagate weights_path from IR FE to the plugin

### PR DESCRIPTION
**Pull request by tkrupa-intel**
This PR is an implementation of an idea requested in this discussion: https://github.com/openvinotoolkit/openvino/pull/25731#discussion_r1802669119.

### CODE REVIEW

**Review by praasz — APPROVED** [review]

File: src/plugins/intel_gpu/src/plugin/plugin.cpp [hunk 1]
```diff
@@ -162,6 +162,22 @@ Plugin::Plugin() {
...
+void Plugin::set_cache_info(const std::shared_ptr<const ov::Model>& model, ExecutionConfig& config) const {
+    // GPU_WEIGHTS_PATH is used for the weightless cache mechanism which is used only with
+    // ov::CacheMode::OPTIMIZE_SIZE setting. Not setting GPU_WEIGHTS_PATH will result in not
+    // using that mechanism.
+    if (config.get_property(ov::cache_mode) != ov::CacheMode::OPTIMIZE_SIZE) {
+        return;
+    }
+
+    const auto& rt_info = model->get_rt_info();
+    auto weights_path = rt_info.find("weights_path");
```
Comment (t-jankowski) [inline src/plugins/intel_gpu/src/plugin/plugin.cpp:174]:
```suggestion
    auto weights_path = rt_info.find("__model_bin_file_path");
```

File: src/frontends/ir/src/input_model.cpp [hunk 2]
```diff
@@ -263,6 +270,8 @@ std::shared_ptr<ov::Model> InputModel::InputModelIRImpl::convert() {
     std::shared_ptr<ov::Model> model;
     visitor.on_attribute("net", model);
     model->get_rt_info()["version"] = int64_t(version);
+    if (!m_weights_path.empty())
+        model->get_rt_info()["weights_path"] = m_weights_path;
```
Comment (t-jankowski) [inline src/frontends/ir/src/input_model.cpp:274]:
```suggestion
        model->get_rt_info()["__model_bin_file_path"] = m_weights_path;
```

File: src/frontends/ir/src/frontend.cpp [hunk 3]
```diff
@@ -168,15 +168,16 @@ InputModel::Ptr FrontEnd::load_impl(const std::vector<ov::Any>& variants) const
...
+            return std::make_shared<InputModel>(*provided_model_stream, weights, create_extensions_map(), weights_path);
         } else if (local_model_stream.is_open()) {
-            auto input_model = std::make_shared<InputModel>(local_model_stream, weights, create_extensions_map());
+            auto input_model =
+                std::make_shared<InputModel>(local_model_stream, weights, create_extensions_map(), weights_path);
             local_model_stream.close();
             return input_model;
         } else if (model_buf) {
-            return std::make_shared<InputModel>(model_buf, weights, create_extensions_map());
+            return std::make_shared<InputModel>(model_buf, weights, create_extensions_map(), weights_path);
```
Comment (t-jankowski) [inline src/frontends/ir/src/frontend.cpp:180]: Regarding `weights_path` use `std::move` or pass by `const &`.

File: src/plugins/intel_gpu/src/plugin/plugin.cpp [hunk 4]
```diff
@@ -162,6 +162,22 @@ Plugin::Plugin() {
     m_compiled_model_runtime_properties["OV_VERSION"] = ov_version.buildNumber;
 }
 
+void Plugin::set_cache_info(const std::shared_ptr<const ov::Model>& model, ExecutionConfig& config) const {
+    // GPU_WEIGHTS_PATH is used for the weightless cache mechanism which is used only with
+    // ov::CacheMode::OPTIMIZE_SIZE setting. Not setting GPU_WEIGHTS_PATH will result in not
+    // using that mechanism.
```
Comment (t-jankowski) [inline src/plugins/intel_gpu/src/plugin/plugin.cpp:168]: I think it should be 'WEIGHTS_PATH', not 'GPU_WEIGHTS_PATH'.
I'd also consider to rename property 'WEIGHTS_PATH' to 'MODEL_BIN_FILE_PATH'.

**Review by t-jankowski — CHANGES_REQUESTED** [review]

File: src/core/src/pass/serialize.cpp [hunk 5]
```diff
@@ -1168,7 +1168,7 @@ void ngfunction_2_ir(pugi::xml_node& netXml,
     pugi::xml_node rt_info_node = netXml.append_child("rt_info");
     for (const auto& it : model.get_rt_info()) {
         // Skip IR version
-        if (it.first == "version")
+        if (it.first == "version" || it.first == "weights_path")
```
[the remark this answers is not in the document]
Comment (tkrupa-intel) [reply src/core/src/pass/serialize.cpp]: Added underscores in the rt_info key but didn't change the name (as agreed in the PM) due to the name already being in use elsewhere.

**Review by t-jankowski — APPROVED** [review]

**Comment by p-durandin** [comment]
build_jenkins

**Review by e-ddykim — APPROVED** [review]
Looks good to me

**Review by p-durandin — APPROVED** [review]

**Pull request merged**